tidy3d/plugins/metrics/README.md

```diff
@@ -0,0 +1,265 @@
+result = expr(mode_coeff1=data1, mode_power1=data2)
+```
+
+#### Important notes on variable evaluation
+
+- **Unnamed Variables**: If you have an unnamed variable (`Variable()`), you must provide exactly one positional argument during evaluation.
+Providing multiple positional arguments will raise a `ValueError`.
+
+- **Named Variables**: For each named variable, you must provide a corresponding keyword argument.
+If a named variable is missing during evaluation, a `ValueError` will be raised.
+
+- **Multiple Unnamed Variables**: If your expression contains multiple unnamed variables, you cannot provide multiple positional arguments.
+You should assign names to the variables and provide their values as keyword arguments instead.
+
+## Developer notes
+
+### Extending metrics
+
+To implement new metrics, follow these steps:
+
+1. **Subclass the `Metric` base class:**
+
+   Create a new class that inherits from `Metric` and implement the required methods.
+
+   ```python
+   from tidy3d.plugins.metrics import Metric
+
+   class CustomMetric(Metric):
+       monitor_name: str
+
+       def evaluate(self, data: SimulationData) -> NumberType:
+           # Implement custom evaluation logic
+           pass
+   ```
+
+2. **Define attributes and methods:**
+
+   - Add any necessary attributes with type annotations.
+   - Implement the `evaluate` method, which defines how the metric is calculated from simulation data.
+
+3. **Update forward references:**
+
+   If your metrics reference other metrics or expressions, ensure you handle forward references appropriately.
+
+### Extending operators
+
+To extend the `metrics` module with additional operators:
+
+1. **Create a new operator class:**
+
+   Subclass `UnaryOperator` or `BinaryOperator` depending on the operator's arity.
+
+   ```python
+   from tidy3d.plugins.metrics import BinaryOperator
+
+   class CustomOperator(BinaryOperator):
+       _symbol = "??"  # Replace with the operator symbol
+       _format = "({left} {symbol} {right})"
+
+       def evaluate(self, x, y):
+           # Implement the operator logic
+           pass
+   ```
+
+2. **Implement required methods:**
+
+   - Define the `_symbol` class variable that represents the operator symbol.
+   - Implement the `evaluate` method with the operator's logic.
+
+3. **Update operator overloads:**
+
+   If you want to enable the use of the operator with standard syntax (e.g., using `@` for matrix multiplication), override the corresponding magic methods in the `Expression` base class.
+
+   ```python
+   class Expression:
+       # Existing methods...
+
+       def __matmul__(self, other):
+           return MatMul(left=self, right=other)
+   ```
+
+4. **Register the operator (if necessary):**
+
+   Ensure the new operator is recognized by the serialization mechanism. If your operator introduces a new type, update any type maps or registries used in the `metrics` module.
```

Comment:
could you elaborate?

Reply:
yeah basically they need to be added to the correct section in `types.py` and if it's a new module, then that module needs to be added in the `__ini__.py`. I'll make this clearer here